media-proxy/src/mesh/conn_rdma.cc

```diff
@@ -0,0 +1,359 @@
+        return Result::error_out_of_memory;
+    }
+
+    // Zero-initialize the entire memory block
+    std::memset(memory_block, 0, total_size);
+
+    // Divide the memory block into individual buffers
+    char *base_ptr = static_cast<char *>(memory_block);
+    for (size_t i = 0; i < capacity; ++i) {
+        void *buf = base_ptr + i * aligned_trx_sz;
+        buffer_queue.push(buf);
+    }
+
+    // Store the base memory block for cleanup
+    buffer_block = memory_block;
+
+    return Result::success;
+}
+
+// Cleanup the buffer queue
+void Rdma::cleanup_queue()
+{
+    if (buffer_block) {
+        // Free the entire memory block
+        std::free(buffer_block);
+        buffer_block = nullptr;
+    }
+    while (!buffer_queue.empty()) {
+        buffer_queue.pop();
+    }
+}
+
+/**
+ * @brief Configures the RDMA connection with the provided parameters.
+ * 
+ * Sets up the RDMA transfer size, connection kind, endpoint configuration,
+ * and queue size based on the input request. Ensures that the kind-direction
+ * pairing is valid (e.g., receiver with RX direction). Updates the connection
+ * state to "configured" upon success.
+ * 
+ * @param ctx The operation context.
+ * @param request The connection parameters including addresses and RDMA arguments.
+ * @param dev_port The device port identifier.
+ * @param dev_handle Reference to the RDMA device handle.
+ * @param kind The type of connection (receiver or transmitter).
+ * @param dir The data transfer direction (RX or TX).
+ * @return Result::success on successful configuration, or an error result if arguments are invalid.
+ */
+Result Rdma::configure(context::Context& ctx, const mcm_conn_param& request,
+                       const std::string& dev_port, libfabric_ctx *& dev_handle)
+{
+
+    trx_sz = request.payload_args.rdma_args.transfer_size;
+    _kind = kind();
+
+    std::memset(&ep_cfg, 0, sizeof(ep_cfg));
+    std::memcpy(&ep_cfg.remote_addr, &request.remote_addr, sizeof(request.remote_addr));
+    std::memcpy(&ep_cfg.local_addr, &request.local_addr, sizeof(request.local_addr));
+    
+    ep_cfg.dir = _kind == Kind::receiver ? direction::RX : direction::TX;
+
+    mDevHandle = dev_handle;
+
+    queue_size = request.payload_args.rdma_args.queue_size;
+
+    set_state(ctx, State::configured);
+    return Result::success;
+}
+
+/**
+ * @brief Establishes an RDMA connection by initializing the device, endpoint, buffer queue,
+ * and starting necessary threads for processing RDMA operations.
+ *
+ * Ensures proper cleanup and state updates on failure. Returns success if all steps
+ * (device initialization, endpoint setup, buffer queue creation, and endpoint configuration)
+ * complete successfully.
+ *
+ * @param ctx The operation context.
+ * @return Result::success on success, or an appropriate error result on failure.
+ */
+Result Rdma::on_establish(context::Context& ctx)
+{
+    if (init) {
```

Comment:
Is it possible that:
- multiple threads will call on_establish() at the same time?
- one thread call on_establish() and the second calls shortly after on_shutdown()?
If yes, we need a better lock then checking `init`

I suspect that multiple threads calling on_establish() is very possible, as currently every connection to proxy creates its own thread

Reply:
No, there is only one decision maker in Media Proxy that calls `establish()` – BridgesManager. The issue won't happen by design.